### ISSUE TITLE
Sharing/coupling ordinal and continuous axes when layering heat map with line graph

### ISSUE THREAD

**Issue by gwdekker** [altair-viz/altair#995]
I am trying to use a heat map as a background for my line graph. The layering, however, seems not to work as intended. Full example included below, grateful for any clarification whether this is a missing feature or whether I am not doing this right.

Code to create sample dataframes:
```
import pandas as pd
import altair as alt
import numpy as np
dfc_binned = pd.DataFrame(data=[[0, '0-0.1%', 305.884748, 3.485558],
    [0, '0.1-1%', 181.717722, 3.259397],
    [0, '1-10%', 0.420775  , 0.624049],
    [0, '10-100%', 0.050000  , 0.000000],
    [1, '0-0.1%', 150.251343, 3.176818],
    [1, '0.1-1%', 1.466669  , 1.166332],
    [1, '1-10%', 1.050000  , 1.021189],
    [1, '10-100%', 1.050000  , 1.021189]], columns = ['Alternative', 'Bin_prob', 'Cost_MNOK', 'Cost_MNOK_log']
)
risk_colors = ['green', 'yellow', 'orange', 'red']
mx = pd.DataFrame(columns=['x', 'y', 'val'], 
                  data=[[0, 0, 'low'], 
                        [0, 1, 'low'],
                        [0, 2, 'low'],
                        [0, 3, 'medium'],
                        [1, 0, 'low'],
                        [1, 1, 'low'],
                        [1, 2, 'medium'],
                        [1, 3, 'high'],
                        [2, 0, 'low'],
                        [2, 1, 'medium'],
                        [2, 2, 'medium'],
                        [2, 3, 'critical'],
                        [3, 0, 'low'],
                        [3, 1, 'medium'],
                        [3, 2, 'high'],
                        [3, 3, 'critical'],
                       ], )
def map_x_to_pct(val):
    if val == 0:
        return '0-0.1%'
    if val == 1:
        return '0.1-1%'
    if val == 2:
        return '1-10%'
    if val == 3:
        return '10-100%'
mx['x2']=mx['x'].apply(map_x_to_pct)
mx['y2']=mx['y']+.5
```

Code to create the line chart works as intended:

```
line = alt.Chart(dfc_binned).mark_line().encode(
    x='Bin_prob',
    y=alt.Y('Cost_MNOK_log', scale=alt.Scale(domain=(0, 4))),
    color='Alternative:N'
)
line
```
Resulting in:
![visualization 3](https://user-images.githubusercontent.com/37177265/42281512-08f9610c-7fa4-11e8-8f7f-8b2be5f9d2fa.png)

code for creating risk heatmap also works as intended:
```
risk = alt.Chart(mx).mark_rect().encode(
    x='x2:O',
    y=alt.Y('y2:O', sort='descending', scale=alt.Scale(domain=(3.5, 2.5, 1.5, 0.5))),
    color=alt.Color('val:O', sort=['low', 'medium', 'high', 'critical'], scale=alt.Scale(range=risk_colors)),
)
risk
```
![visualization 4](https://user-images.githubusercontent.com/37177265/42281546-22a0ee0e-7fa4-11e8-8d8c-3019ca710e41.png)

Now, I would like to overlay these two on top of each other, where the axes would line up and the original colors and legends remain intact. The easiest way I found was using `risk+line`, I also tried using `alt.layer(risk, line).resolve_scale(y='shared').resolve_legend(color='independent')`, but the result is in any case pretty much the same and clearly not as intended: 
![visualization 5](https://user-images.githubusercontent.com/37177265/42281718-b32f8a5c-7fa4-11e8-81f7-116e063e25c6.png)

Thanks in advance for any clarification!

**Comment by tweakimp** [comment]
I think I got it:

```python
mx_transfer = mx[["x2", "y2", "val"]]
mx_transfer.columns = ["Bin_prob", "Cost_MNOK_log", "Alternative"]
dfc_binned_transfer = dfc_binned[["Bin_prob", "Cost_MNOK_log", "Alternative"]]
df = mx_transfer.append(dfc_binned_transfer)

lines = (
    alt.Chart(df)
    .mark_line()
    .encode(alt.X("Bin_prob:N"), alt.Y("Cost_MNOK_log"), color="Alternative:N")
    .transform_filter(alt.FieldOneOfPredicate(field="Alternative", oneOf=[0, 1]))
)
rects = (
    alt.Chart(df)
    .mark_rect()
    .encode(
        alt.X("Bin_prob:N"),
        alt.Y("Cost_MNOK_log:Q", bin=alt.Bin(maxbins=4)),
        color=alt.Color(
            "Alternative:N", sort=["low", "medium", "high", "critical"], scale=alt.Scale(range=risk_colors)
        ),
    )
    .transform_filter(alt.FieldOneOfPredicate(field="Alternative", oneOf=["low", "medium", "high", "critical"]))
)


chart = alt.layer(rects, lines).properties(
    width=400, height=400, resolve=alt.Resolve(scale=alt.LegendResolveMap(color=alt.ResolveMode("independent")))
)
```

![mytry](https://i.imgur.com/pxy2gQr.png)

First I had to build one dataframe with exactly that data which is in the graph. This made it easier to look build it. Also putting all data which would eventually be on one axis be on the same column made it easier to fix the problem with the data types.
The keys in building the chart were to use `bin` for the y axis to make altair use the same axis for both the heatmap and the lines and the `resolve` attribute in the chart properties. You can now change the titles of the legend to and the size of the rectangles so that there are no gaps.

**Comment by gwdekker** [comment]
Thanks a lot for the help! I picked over your answer and tried to distill the parts that made the difference. Turns out I did not need to combine into one dataframe or use transform_filter operations to get the result I wanted. The following elements in your answer solved my problem:

**1. Solving the axis scaling: using fixed width and height.**
Using the subcharts as defined in my first post and combining them with a fixed width and height forces the axes to match:
```
alt.layer(risk, line).properties(
    width=400, height=400)
```
![visualization](https://user-images.githubusercontent.com/37177265/42305894-473cd500-802d-11e8-8ca1-c46cabf67ec0.png)

**2. Solving the color coding: `resolve_scale`, not `resolve_legend`**
In my first tries, I was using `resolve_legend` while in fact I was looking for `resolve_scale` as it is not the legend but the scale I was trying to decouple. Fixing this gives me the chart I was looking for:
```
alt.layer(risk, line).properties(
    width=400,
    height=400).resolve_scale(color="independent")
```
![visualization 1](https://user-images.githubusercontent.com/37177265/42306151-39d4d632-802e-11e8-8fba-ff9ecee24d66.png)

Thanks again!
